src/context/GlobalContext.tsx

```diff
@@ -0,0 +1,62 @@
+import React, {
+  useContext,
+  useReducer,
+  useEffect,
+  createContext,
+  useMemo,
+} from "react";
+import {
+  ContextActions,
+  GlobalState,
+  Types,
+} from "src/context/context.interface";
+import * as ActionCreators from "src/context/ActionCreators";
+import { rootReducer } from "src/context/reducer";
+import { AuthFormData } from "src/common/interface";
+import { auth } from "src/modules/AuthService";
+import { User } from "src/modules/User";
+
+const initialState: GlobalState = {
+  loading: true,
+};
+const GlobalContext = createContext<{
+  state: GlobalState;
+  actions?: ContextActions;
+}>({
+  state: initialState,
+});
+
+export function useAuth() {
+  return useContext(GlobalContext);
+}
+
+export const GlobalProvider: React.FC = ({ children }) => {
+  const [state, dispatch] = useReducer(rootReducer, initialState);
+
+  const actions = useMemo<ContextActions>(
+    () => ({
+      signUp: (payload: AuthFormData) =>
+        ActionCreators.register(dispatch, payload),
+      signIn: (payload: AuthFormData) =>
+        ActionCreators.login(dispatch, payload),
+      signOut: () => ActionCreators.logout(dispatch),
+    }),
+    []
+  );
+
+  useEffect(() => {
+    const subscribeAuthChange = auth.onAuthStateChanged((user) => {
+      dispatch({
+        type: Types.SET_USER,
+        payload: user ? new User(user) : undefined,
```

Comment:
Only the Auth module is allowed to create a `User` object

Reply:
if we reload our component we have to create `User` object because only `signIn` and `signUp` return with `User` object. and we are not storing it in local storage.

Reply:
Please see my comment inside `onAuthStateChanged` implementation.